### ISSUE TITLE
Move settings to pref domain

### ISSUE THREAD

**Issue by clburlison** [clburlison/Munki-s3Repo-Plugin#2]
We'll override boto3 defaults with a preference domain so we can extend. Once we add extra options we will hit limits on customization. This will also allow support for minio, custom acls, etc.

**Comment by clburlison** [comment]
For multiple reasons I don't think I'll be using CFPrefs for this.
* It's impossible to have nested dictionaries with `defaults write`
* No valid reason that I can think of for why this pref file should not be under $HOME
* I want to keep the barrier of entry low
* Could solve with a profile but again I'm leaning against that as it's yet one more step
* Munki admins will rarely make changes to this once things are up and running

My current working theory for how the pref file will look:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>default</key>
	<dict>
		<key>ExtraArgs</key>
		<dict>
			<key>ACL</key>
			<string>public-read</string>
			<key>GrantFullControl</key>
			<string>id=asdfasdfewrqwesdfgsdfgxcvxcb</string>
			<key>StorageClass</key>
			<string>REDUCED_REDUNDANCY</string>
		</dict>
		<key>Metadata</key>
		<dict>
			<key>Cache-Control</key>
			<integer>86400</integer>
		</dict>
		<key>aws_access_key_id</key>
		<string>12123123123123</string>
		<key>aws_secret_access_key</key>
		<string>asdfasdfasdfasdasdfasdfasdfasdf</string>
		<key>bucket</key>
		<string>clburlison-munkirepo</string>
		<key>region</key>
		<string>us-east-1</string>
	</dict>
</dict>
</plist>
```

And you can set these values with:

```bash
/usr/libexec/PlistBuddy -c "clear dict" \
-c "add :default dict" \
-c "add :default:aws_access_key_id string '12123123123123'" \
-c "add :default:aws_secret_access_key string 'asdfasdfasdfasdasdfasdfasdfasdf'" \
-c "add :default:bucket string 'clburlison-munkirepo'" \
-c "add :default:region string 'us-east-1'" \
-c "add :default:ExtraArgs dict" \
-c "add :default:ExtraArgs:ACL string 'public-read'" \
-c "add :default:ExtraArgs:GrantFullControl string 'id=asdfasdfewrqwesdfgsdfgxcvxcb'" \
-c "add :default:ExtraArgs:StorageClass string 'REDUCED_REDUNDANCY'" \
-c "add :default:Metadata dict" \
-c "add :default:Metadata:Cache-Control integer 86400" \
~/Library/Preferences/com.clburlison.munki.s3Repo.plist
```

The reason for having the preferences inside of their own dictionary is so I can allow for multiple "aws profiles". Since we're overriding Boto's default preference list the admin will have to recreate these profiles but it will allow us to have a `default`, `production`, `local`, etc and be able to easily switch between them with:

```bash
export AWS_PROFILE=local
```

[`ExtraArgs`](http://boto3.readthedocs.io/en/latest/reference/customizations/s3.html#boto3.s3.transfer.S3Transfer.ALLOWED_UPLOAD_ARGS) & `Metadata` allow us to straight extend the boto library without additional code complexity. Since all of these values are passed to the upload methods the admin can pass any arbitrary data they want.

This isn't implemented yet but the major design decisions have been made and the code shouldn't be hard to write now. 

@rickheil, @erikng thoughts on this implementation?

**Comment by erikng** [comment]
I would personally prefer to use profiles. Is the worry that admins don't know how to write a python script setting their values vs defaults command?

Chances are the people using this plugin know how to interact with CFPreferences.

The structure looks good.
________________________________

**Comment by clburlison** [comment]
_it will be simple he thought_

**Comment by clburlison** [comment]
Alright, both of y'all got your way. I ran into a minor speed bump that was "fun" however design and implementation-wise I am happy with the end result. 

I've modified the README to match the new setup process. s3Repo plugin is now using CFPrefs for configuration values. I've also added a helper script, `prefSetter.py`, to assist admins with setting the values properly. The only change from my above design was the Metadata key/value had to be inside of the ExtraArgs. 

I'll leave this issue open for a bit and if anyone runs into issues let me know.
